edx/edx-exams
# LTI Procotoring Service Specification Proof-Of-Concept: LTI Launch

**JIRA:** https://openedx.atlassian.net/browse/MST-1260

### Setup Steps

1. Clone the [xblock-lti-consumer](https://github.com/openedx/xblock-lti-consumer) repository into your `{workspace}/src` directory. `{workspace}` should be the directory in which you have your other edX git repositories and should be the same directory your `devstack` repository is located in.
2. Check out the `mroytman/MST-1260-special-exams-LTI-launch-POC` branch locally.
3. In `edx-exams`, create a virtual environment.
4. Activate the virtual environment.
5. Install requirements with `make requirements`.
6. Run `make docker_build` to build the Docker image from the `Dockerfile`.
7. Run `make dev.provision` to provision and start the container.
8. Shell into the container by running `make app-shell`. Once inside the shell, run `pip install -e /edx/src/xblock-lti-consumer`. This will install the branch you checked out locally as an editable installation. Alternatively, you can just avoid the step of checking out the `xblock-lti-consumer` repository if you specify a commit-ish to install.

### Testing Steps

If you want to test this implementation against the IMS Proctoring Specification testing tool, follow these steps.

1. Sign up for an IMS account as an employee of 2U [here](https://site.imsglobal.org/user/register), if you are not already a member. The approval process should take around a day.
2. You will need to set up port forwarding so that your running Docker container can connect with the IMS testing tool. You can use `ngrok` for this or any other tool you like. `ngrok` has comprehensive documentation for how to accomplish this.
3. Once your membership has been approved, go to the testing tool [here](https://certification.imsglobal.org/proctoring/launch).
4. In the testing tool, click `Get Started`.
5. Enter your name, email, and organization and click `Confirm`.
6.  Click `Create` under `Create New Product`.
7. Enter a `Product` and `Version` and click `Confirm`. These values can be anything for the purposes of testing.
8. Click `Select` under `Assessment Platform`.
9. Click `Begin Configuration`.
10. Open up the Django admin by visiting `localhost:8000/admin` and logging in with the default edX development login credentials.
11. Under `LTI_CONSUMER > Lti configuration`, click `Add`.
12. Enter the following configuration:
13.1. `Version`: `LTI 1.3 (with LTI Advantage Support)`
13.2. `Config store`: `Configuration Stored on this model`.
14. From the testing tool, copy the following configuration into the `Lti configuration` model in the Django admin.
14.1. `Tool End-Point to Receive Launches` > `Lti 1p3 launch url`
14.2. `OIDC Login Initiation URL` > `Lti 1p3 oidc url`
15. Click `Next`.
16. Enter the following configuration in the testing tool. In this example, `{root}` refers to the URL that is being forwarded. For example, `https://9c11-73-238-153-173.ngrok.io`.
16.1. `Testing iss Value`: `http://localhost:18000`
16.2. `OIDC Auth URL`: `{root}/api/v1/oidc_auth_request`
16.3. `Platform Well-Known/JWKS URL`: `{root}/api/v1/keyset`
16.4. `Client Id (Assigned to Cert Suite)` > Enter the `Lti 1p3 client id` from the `Lti configuration` model in the Django admin.
17. Click `Next`.
18. Enter `1` for the deployment ID and click `Next`.
19. From the testing tool, copy the`Proctoring Tool JWKS URL` into the `Lti configuration` model in the Django admin in the `Lti 1p3 tool keyset url` field.
20. Click `Next`.
21. Check `Pause on OIDC Exchanges to Inspect Values` and/or `Pause Prior to JWT Messages to Inspect Values` depending on whether you want interstitials before Tool requests or responses to the Assessment Platform.
22. Visit `{root}/api/v1/lti_start_proctoring` to kick off the flow.
